### ISSUE TITLE
Docker exiting with environment variable not set error

### ISSUE THREAD

**Issue by madhusudhanarya** [mwmason/standardnotes-onecontainer#2]
Thank you very much for putting the entire standard notes containers into one. I have been trying to get their docker up and running without any success thats when I stumbled upon your repo.

I was able to get the docker image created and able to start the container. But, the container is exiting with below error.

```
Set app_name in your newrelic.js file or set environment variable
NEW_RELIC_APP_NAME. Not starting!

/var/www/syncing-server-js/dist/src/Bootstrap/Env.js:23
            throw new Error(`Environment variable ${key} not set`);
            ^
Error: Environment variable DB_HOST not set
```

Below is the snippet from my env file where i have set the options.

```
root@unraid-server# egrep -i 'app_name|db_host' .env 
DB_HOST=192.168.0.10
NEW_RELIC_APP_NAME=standard-notes
```

can you please let me know what i could be doing wrong here?

**Comment by mwmason** [comment]
Hi @madhusudhanarya,

Glad you're finding this useful - just be aware that my build doesn't include the new mono-repo that SN has moved to - and the version of code is from July-ish, and it doesn't include Files.  I'm running that older version in my production instance, and my clients (android and AppImage) are functioning okay.  

I plan to update to their mono soon - but haven't had the cycles recently.

Regarding your error - you still need to define all the required environment variables - each micro service has a respective run script (e.g. for syncing-server: services/syncsvr/run) which contains the s6 command with-contenv.  This command will include all your environment variables that were previously defined in your container.  You can also modify each run to define service specific variables; in my example, I defined syncing-server's PORT to 3000 using:  'export PORT 3000' .  So you could also define your DB_HOST in each service, if you prefer this method.

Let me know if that makes sense, or if you have further questions.

**Comment by madhusudhanarya** [comment]
Thank you @mwmason. I was able to spin up the container by passing `--env-file` that points to all ENV variables.

```
root@unraid-server# docker run -d --env-file=./env --name standard-notes standard-notes
afe92c0c508471cd25f37d274839ec5cb1e327f35bac971fbde8afbb1c8a9cd6
```

Now, i see a different error message in the docker logs. Below is the extract from the log.

```
/var/www/syncing-server-js/node_modules/redis-parser/lib/parser.js:179
    return new ReplyError(string)

           ^
ReplyError: ERR DB index is out of range
    at parseError (/var/www/syncing-server-js/node_modules/redis-parser/lib/parser.js:179:12)
    at parseType (/var/www/syncing-server-js/node_modules/redis-parser/lib/parser.js:302:14) {
  command: { name: 'select', args: [ '20' ] }
}
```

can you please suggest what could be wrong here or where i need to check?

**Comment by mwmason** [comment]
Can you paste the log from stdout?  I want to see what s6 is successfully starting.

**Comment by madhusudhanarya** [comment]
Please find the attached log that i downloaded from the docker container. Few hundred lines from the bottom shows the services starting before raising the exceptions.

[_standard-notes_logs (3).txt](https://github.com/mwmason/standardnotes-onecontainer/files/10133133/_standard-notes_logs.3.txt)

**Comment by mwmason** [comment]
Ok, so your redis instance started, and doesn't exit immediately, so that's good.

Can you connect to your container via cli and issue:

redis-cli -h localhost -p 6379 ping

You should get a PONG as a response.

**Comment by madhusudhanarya** [comment]
I get the expected response.

![image](https://user-images.githubusercontent.com/31273704/205097097-46cce4d0-44e2-4101-99e8-37fb86755a46.png)

**Comment by mwmason** [comment]
Hmm... and you're REDIS_URL environment variable is:

REDIS_URL=redis://localhost:6379

?

**Comment by mwmason** [comment]
Yeah - I think the redis error is a red herring.  I just looked at the parser.js file and it's expecting an integer.  

Can you connect again to the container and look for a log file for syncsvr in /var/log/syncsvr .  I'm thinking the "ERR_DB" is really a database connection issue.  If syncsvr really connected to your database you should see new tables created.

**Comment by madhusudhanarya** [comment]
Below is the output that i can see in the log file.

```
/var/log/syncsvr # tail current 
2022-12-03 05:56:35.238360661  query: SELECT * FROM `snotes`.`migrations` `migrations` ORDER BY `id` DESC
2022-12-03 05:56:35.239553970  No migrations are pending
2022-12-03 05:56:35.257044040  {"message":"Server started on port 3000","level":"info"}
2022-12-03 05:56:36.202907225  {"v":0,"level":30,"name":"newrelic_bootstrap","hostname":"asfah8789asdf","pid":1441,"time":"2022-12-03T05:56:35.590Z","msg":"NEW_RELIC_NO_CONFIG_FILE set, deferring to environment variables."}
2022-12-03 05:56:36.218136976  query: SELECT VERSION() AS `version`
2022-12-03 05:56:36.233542077  INFO: All classes found using provided glob pattern "/var/www/syncing-server-js/dist/migrations/*.js" : "/var/www/syncing-server-js/dist/migrations/1606470249552-init_database.js,/var/www/syncing-server-js/dist/migrations/1617615657558-add_extension_settings.js,/var/www/syncing-server-js/dist/migrations/1629964808297-drop_unused_indexes.js,/var/www/syncing-server-js/dist/migrations/1630318893601-refactor_calculating_integrity_hash.js,/var/www/syncing-server-js/dist/migrations/1630417724617-restrict_content_type.js,/var/www/syncing-server-js/dist/migrations/1631529502150-add_revision_for_duplicated_items.js,/var/www/syncing-server-js/dist/migrations/1631530260504-drop_item_revisions_joining_table.js,/var/www/syncing-server-js/dist/migrations/1632219307742-cleanup_orphan_items_and_revisions.js,/var/www/syncing-server-js/dist/migrations/1632221263106-add_revisions_items_relation.js,/var/www/syncing-server-js/dist/migrations/1637738491169-add_item_content_size.js,/var/www/syncing-server-js/dist/migrations/1639134926025-remove_extension_settings.js,/var/www/syncing-server-js/dist/migrations/1642073387521-remove_sf_extension_items.js,/var/www/syncing-server-js/dist/migrations/1647501696205-remove_user_agent.js,/var/www/syncing-server-js/dist/migrations/1654518291191-add_updated_with_session.js"
2022-12-03 05:56:36.241294329  query: SELECT * FROM `INFORMATION_SCHEMA`.`COLUMNS` WHERE `TABLE_SCHEMA` = 'snotes' AND `TABLE_NAME` = 'migrations'
2022-12-03 05:56:36.245065276  query: SELECT * FROM `snotes`.`migrations` `migrations` ORDER BY `id` DESC
2022-12-03 05:56:36.246277465  No migrations are pending
2022-12-03 05:56:36.262716485  {"message":"Server started on port 3000","level":"info"}
```

I can see the DB that i have set for standard notes is online and has tables

```
sh-4.4# 
sh-4.4# mysql
ERROR 1045 (28000): Access denied for user 'root'@'localhost' (using password: NO)
sh-4.4# mysql -u snotes-user -p
Enter password: 
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 231
Server version: 8.0.31 MySQL Community Server - GPL

Copyright (c) 2000, 2022, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> show databases;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| performance_schema |
| snotes             |
+--------------------+
3 rows in set (0.01 sec)

mysql> use snotes;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
mysql> show tables;
+---------------------------------+
| Tables_in_snotes                |
+---------------------------------+
| analytics_entities              |
| items                           |
| migrations                      |
| offline_settings                |
| offline_user_roles              |
| offline_user_subscriptions      |
| permissions                     |
| revisions                       |
| revoked_sessions                |
| role_permissions                |
| roles                           |
| sessions                        |
| settings                        |
| shared_subscription_invitations |
| subscription_settings           |
| user_roles                      |
| user_subscriptions              |
| users                           |
+---------------------------------+
18 rows in set (0.00 sec)
```

**Comment by mwmason** [comment]
I'm wondering if my comment in the env.sample is contaminating the REDIS_URL value:

REDIS_URL=redis://localhost:6379 #redis will run on localhost default port 6379

Can you echo $REDIS_URL from a command prompt - let's just rule it out.  Has to be something simple because I've never seen an error like this.

**Comment by madhusudhanarya** [comment]
Below is the output

![image](https://user-images.githubusercontent.com/31273704/205430174-d313a460-76e3-4f58-b818-127e0e618621.png)

**Comment by madhusudhanarya** [comment]
After removing the comment string, i can see the container has started. But i am not able to access it using http://IP:3002. I don't see any port mapped to this container.

![image](https://user-images.githubusercontent.com/31273704/205430457-b1e18561-d4d1-4cfb-bb73-234175279f94.png)

Any changes need in env file?

**Comment by mwmason** [comment]
Yeah - you have to map the external port with your docker command.

**Comment by mwmason** [comment]
There were similar comments on AUTH_SERVER_URL, SYNCING_SERVER_JS_URL and SYNCING_SERVER_URL - make sure to remove those as well.  I've already updated the env.sample.

**Comment by madhusudhanarya** [comment]
Thank you very much @mwmason  for assisting me in getting your image running. I am able to start and access the API endpoint after exposing the port.

Slightly unrelated topic. Were you able to connect to self hosted standardnotes instance using either standardnotes-web or android client? I am trying to get their WEB application configured (https://github.com/standardnotes/app/tree/main/packages/web) but failing to do so with below error. I am pointing to the instance that i deployed using your image. Tried changing from localhost to the ipaddress, but no luck

![image](https://user-images.githubusercontent.com/31273704/205446397-88282ebd-d150-4f1e-bb42-352c83853621.png)

![image](https://user-images.githubusercontent.com/31273704/205446326-c38a448a-45e1-413a-889e-bc4c4e6def8f.png)

Tried all possible things that i can image but no luck. If you have any pointer, please do share.

**Comment by mwmason** [comment]
Excellent - apologies I didn't realize my comments would affect the deploy.

I believe the Android app will only work with a valid public SSL certificate (Google enforces that), so my production instance functions with Android and SN's public website.  I doubt you'll be able to connect their website to your localhost instance unless you reverse proxy it to the web.   

When I'm working with my dev/test instance, I only use AppImage.

**Comment by mwmason** [comment]
Corrected in [Release 2.1.2](https://github.com/mwmason/standardnotes-onecontainer/blob/4a895a7e0ed3d3d51e2c804821e2dff3e8a4ecb4/env.sample)
